service/storage_service.cc

```diff
@@ -2259,6 +2259,20 @@ future<> storage_service::do_update_topology_with_local_metadata(raft::server& r
             break;
         }
 
+        // It might happen that, in the previous run, the node commits a command
+        // that adds support for a feature, crashes before applying it and now
+        // it is not safe to disable support for it. If there is an attempt to
+        // downgrade the node then `enable_features_on_startup` called much
+        // earlier won't catch it, we only can do it here after performing
+        // a read barrier - so we repeat it here.
+        //
+        // Fortunately, there is no risk that this feature was marked as enabled
+        // because it requires that the current node responded to a barrier
+        // request - which will fail in this situation.
+        const auto& enabled_features = _topology_state_machine._topology.features.enabled_features;
```

Comment:
This shouldn't be the features from the topology because we may not wrote them yet. It should be features provided on the command line.

Reply:
I don't understand. `check_features` accepts two arguments: a list of features considered enabled by the cluster, and a list of features unsafe to disable, according to the cluster (i.e. features supported by everybody but not yet enabled).

Why would we pass supported features as the first argument? If you look at the `check_features` implementation, it calculates the `supported_feature_set()` based on the node's configuration/CLI arguments and checks them against arguments, so this looks correct to me.

Reply:
OK. I misunderstood what `check_features` receives as a parameters. I thought the first one is currently supported feature by this node. But then why have two parameters to the function? All features in both sets should be present so why not just merge them before passing to the function? Is it just to get different error message?

Reply:
Yes, the function produces a different message depending on whether a feature is missing from the first or the second set. I can see why it looks confusing, though I don't have an idea how to make it look better.